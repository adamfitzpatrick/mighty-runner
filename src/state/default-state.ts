import { Character, PersonalData, CharacterIdentifier, Attributes, Effect, GearItem, Pic } from '@models'

export interface AppState {
  characters: Character[] | null
  activeCharacter: CharacterIdentifier | null
  pic: Pic | null
  personalData: PersonalData | null
  attributes: Attributes | null
  gear: GearItem[] | null
  effects: Effect[] | null
  apiError: boolean
}

export const defaultState: AppState = {
  characters: null,
  pic: null,
  activeCharacter: null,
  personalData: null,
  attributes: null,
  gear: null,
  effects: null,
  apiError: false
}

export default defaultState
