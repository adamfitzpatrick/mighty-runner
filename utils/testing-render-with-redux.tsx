import * as React from 'react'
import {
  createStore,
  combineReducers,
  AnyAction,
  Store,
  Dispatch,
  applyMiddleware
} from 'redux'
import { Provider } from 'react-redux'
import { render, RenderResult } from '@testing-library/react'
import { AppState, defaultState } from '../src/state/default-state'
import {
  characters,
  activeCharacter,
  personalData,
  attributes,
  gear,
  effects
} from '../src/state/reducers'

export type RenderWithRedux =
  RenderResult &
  {
    store: Store<AppState, AnyAction>,
    dispatchSpy: jest.Mock<Dispatch<AnyAction>>
  }

const reducer = combineReducers({
  characters,
  activeCharacter,
  personalData,
  attributes,
  gear,
  effects
})

const getSpyMiddleware = (dispatchSpy: jest.Mock<Dispatch<AnyAction>>) => () => {
  return (next: Dispatch<AnyAction>) => (action: AnyAction) => {
    dispatchSpy(action)
    next(action)
  }
}

export function renderWithRedux (
  ui: JSX.Element,
  initialState: AppState,
): RenderWithRedux {
  const dispatchSpy: jest.Mock<Dispatch<AnyAction>> = jest.fn()
  const spyMiddleware = getSpyMiddleware(dispatchSpy)
  const store = createStore(reducer, initialState, applyMiddleware(spyMiddleware))

  return {
    ...render(<Provider store={store}>{ ui }</Provider>),
    store,
    dispatchSpy
  }
}
