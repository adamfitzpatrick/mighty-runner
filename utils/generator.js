const fs = require('fs')
const path = require('path')
const yargs = require('yargs')

if (!yargs.argv.container && !yargs.argv.component) {
  throw new Error('no component or container name provided')
}

const container = yargs.argv.container && yargs.argv.container.replace(/^[a-z]/, a => a.toUpperCase())
const component = yargs.argv.component && yargs.argv.component.replace(/^[a-z]/, a => a.toUpperCase())

let thing = yargs.argv.component
if (container) {
  thing = container
}
const lower = thing.replace(/^[A-Z]/, a => a.toLowerCase()).replace(/[A-Z]/g, a => `-${a.toLowerCase()}`)
thing = thing.replace(/^[a-z]/, a => a.toUpperCase()).replace(/-([A-Za-z0-9])/g, a => a.slice(1).toUpperCase())
scssClass = thing.replace(/^[A-Z]/, a => a.toLowerCase())

const thingTs = `import * as React from 'react'

import * as styles from './${lower}.scss'

interface Props {}

export default function ${thing} (props: Props) {
  return <div>${thing}</div>
}
`

const thingTestTs = `import * as React from 'react'
import { cleanup, render } from '@testing-library/react'
import { renderWithRedux, RenderWithRedux } from '../../../utils/testing-render-with-redux'
import ${thing} from '.'

describe('${thing} ${ container ? 'container' : 'component'}', () => {
  test('should do nothing', () => {
    expect(true).toEqual(true)
  })
})
`

const indexTs = `export { default } from './${lower}'
`

const scss = `.${scssClass} {
  display: block;
}
`

const scssD = `export const ${scssClass}: string
`

let dest = path.resolve(process.cwd(), 'src', 'components', lower)
if (container) {
  dest = path.resolve(process.cwd(), 'src', 'containers', lower)
}

fs.mkdirSync(dest)
fs.writeFileSync(path.resolve(dest, `${lower}.tsx`), thingTs)
fs.writeFileSync(path.resolve(dest, `${lower}.test.tsx`), thingTestTs)
fs.writeFileSync(path.resolve(dest, 'index.ts'), indexTs)
fs.writeFileSync(path.resolve(dest, `${lower}.scss`), scss)
fs.writeFileSync(path.resolve(dest, `${lower}.scss.d.ts`), scssD)
