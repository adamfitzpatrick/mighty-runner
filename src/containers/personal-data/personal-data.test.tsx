import '@testing-library/jest-dom/extend-expect'
import * as React from 'react'
import { Provider } from 'react-redux'
import { cleanup, fireEvent } from '@testing-library/react'
import { renderWithRedux, RenderWithRedux } from '../../../utils/testing-render-with-redux'
import * as Models from '@models'
import PersonalData from '.'
import { AppState } from '@state/default-state'

describe('PersonalData container', () => {
  let personalData: Models.PersonalData
  let appState: AppState
  let sut: RenderWithRedux

  beforeEach(() => {
    personalData = {
      name: 'n',
      role: 'r',
      description: 'd',
      metatype: 'm',
      ethnicity: 'e',
      age: 32,
      gender: 'g',
      weight: 'w',
      height: 'h',
      karma: 1,
      notoriety: 2,
      streetCred: 3,
      publicAwareness: 4,
      totalKarma: 5,
      nuyen: 10000,
      lifestyle: 'Low'
    }
    appState = {
      characters: null,
      activeCharacter: null,
      pic: null,
      personalData,
      attributes: null,
      gear: null,
      effects: null,
      apiError: false
    }
    sut = renderWithRedux(<PersonalData />, appState)
  })

  test('should render correctly', () => {
    expect(sut.container.innerHTML).toContain('<div data-testid=\"personal-data.container\"')
  })

  test('should render correctly when there is no personal data to show', () => {
    appState.personalData = null
    sut = renderWithRedux(<PersonalData />, appState)
    expect(sut.container.innerHTML).toBe('')
  })

  test('should render with an empty string in place of an undefined data field', () => {
    delete appState.personalData!.name
    cleanup()
    sut = renderWithRedux(<PersonalData />, appState)
    expect((sut.getByTestId('Name.input.component')as HTMLInputElement).value).toEqual('')
  })

  test('should update the application personalData state on an input change', () => {
    fireEvent.change(sut.getByTestId('Name.input.component'), { target: { value: 'change' } })
    personalData.name = 'change'
    expect(sut.dispatchSpy).toHaveBeenCalledWith({
      type: 'SET_PERSONAL_DATA',
      payload: personalData
    })
  })

  test('should trigger a save of the personalData on input field blur', () => {
    fireEvent.blur(sut.getByTestId('Name.input.component'))
    expect(sut.dispatchSpy).toHaveBeenCalledWith({ type: 'SAVE_CHARACTER' })
  })
})
