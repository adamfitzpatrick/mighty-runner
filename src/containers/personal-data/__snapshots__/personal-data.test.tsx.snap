// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonalData container should render correctly 1`] = `"<div data-testid=\\"personal-data.container\\"><label><span>Name</span><input data-testid=\\"Name.input.component\\" type=\\"text\\" value=\\"n\\"></label><label><span>Metatype</span><input data-testid=\\"Metatype.input.component\\" type=\\"text\\" value=\\"m\\"></label><label><span>Ethnicity</span><input data-testid=\\"Ethnicity.input.component\\" type=\\"text\\" value=\\"e\\"></label><label><span>Gender</span><input data-testid=\\"Gender.input.component\\" type=\\"text\\" value=\\"g\\"></label><label><span>Age</span><input data-testid=\\"Age.input.component\\" type=\\"text\\" value=\\"32\\"></label><label><span>Height</span><input data-testid=\\"Height.input.component\\" type=\\"text\\" value=\\"h\\"></label><label><span>Weight</span><input data-testid=\\"Weight.input.component\\" type=\\"text\\" value=\\"w\\"></label><label><span>Karma</span><input data-testid=\\"Karma.input.component\\" type=\\"text\\" value=\\"1\\"></label><label><span>Street Cred</span><input data-testid=\\"Street Cred.input.component\\" type=\\"text\\" value=\\"3\\"></label><label><span>Notoriety</span><input data-testid=\\"Notoriety.input.component\\" type=\\"text\\" value=\\"2\\"></label><label><span>Public Awareness</span><input data-testid=\\"Public Awareness.input.component\\" type=\\"text\\" value=\\"4\\"></label><label><span>Total Karma</span><input data-testid=\\"Total Karma.input.component\\" type=\\"text\\" value=\\"5\\"></label></div>"`;

exports[`PersonalData container should render correctly when there is no personal data to show 1`] = `""`;
